map.js

```diff
@@ -103,20 +101,38 @@ function drawBaseMap(){
 function initOverlay(){
     overlay = {
         locations : [],
+        tsp_locations : [],
         nirnroots : [],
+        tsp_nirnroots : [],
         lastZoomLevel : zoomLevel
     }
 
     runOnTree(jsondata.location, function(loc){
         overlay.locations.push(new MapIcon(loc));
+        //takes only what we need for TSP: x, y, and order. 
+            //should we make a seperate "TSP object?" or just leave this as is?
+        overlay.tsp_locations.push({x:loc.x, y:loc.y, tsp:loc.tsp, cell:{x:loc.x, y:loc.y}});
     });
 
-    runOnTree(jsondata.nirnroot, function(loc){
-        if(loc.cell == "Outdoors"){
-            overlay.nirnroots.push(new MapIcon(loc));
+    runOnTree(jsondata.nirnroot, function(nirn){
+        if(nirn.cell == "Outdoors"){
+            overlay.nirnroots.push(new MapIcon(nirn));
+
+            //takes only what we need for TSP: x, y, and order.
+            overlay.tsp_nirnroots.push({x:nirn.x, y:nirn.y, tsp:nirn.tsp, cell:{x:nirn.x, y:nirn.y}});
         }
     });
+
+    //init, calculate, and save TSP array.
+    overlay.tsp_locations.sort((a, b) => a.tsp - b.tsp);
+    overlay.tsp_nirnroots.sort((a, b) => a.tsp - b.tsp);
+
+    //remove the tsp:-1 locations, since they're not calculated into a tsp yet.
+    while(overlay.tsp_locations[0].tsp == -1)overlay.tsp_locations.shift();
+    while(overlay.tsp_nirnroots[0].tsp == -1)overlay.tsp_nirnroots.shift();
+    recalculateTSP();
```

Comment:
recalculateTSP() works kind of like recalculate bounding box, where it calculates world to map space cords on zoom change.

Reply:
We could check the tsp_locations for -1 (or null) when we insert them into the  array, that way we don't have to check here

Reply:
Done!